cpp/src/arrow/acero/source_node_test.cc

```diff
@@ -0,0 +1,109 @@
+
+  int num_pass;
+};
+
+struct PauseThenStopNode : public MapNode {
+  static constexpr const char* kKindName = "PauseThenStopNode";
+  static constexpr const char* kFactoryName = "pause_then_stop";
+
+  static void Register() {
+    auto exec_reg = default_exec_factory_registry();
+    if (!exec_reg->GetFactory(kFactoryName).ok()) {
+      ASSERT_OK(exec_reg->AddFactory(kFactoryName, PauseThenStopNode::Make));
+    }
+  }
+
+  PauseThenStopNode(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                    std::shared_ptr<Schema> output_schema,
+                    const PauseThenStopNodeOptions& options)
+      : MapNode(plan, inputs, output_schema), num_pass(options.num_pass) {}
+
+  static Result<ExecNode*> Make(ExecPlan* plan, std::vector<ExecNode*> inputs,
+                                const ExecNodeOptions& options) {
+    RETURN_NOT_OK(ValidateExecNodeInputs(plan, inputs, 1, kKindName));
+    auto bp_options = static_cast<const PauseThenStopNodeOptions&>(options);
+    if (bp_options.num_pass < 2) {
+      return Status::Invalid("num_pass must be at least 2");
+    }
+    return plan->EmplaceNode<PauseThenStopNode>(plan, inputs, inputs[0]->output_schema(),
+                                                bp_options);
+  }
+
+  const char* kind_name() const override { return kKindName; }
+  Result<ExecBatch> ProcessBatch(ExecBatch batch) override {
+    if (num_pass == 2) {
+      inputs()[0]->PauseProducing(this, 1);
+    } else if (num_pass == 1) {
+      ARROW_RETURN_NOT_OK(inputs()[0]->StopProducing());
+    }
+    if (num_pass > 0) --num_pass;
+    return batch;
+  }
+
+  int num_pass;
+};
+
+TEST(SourceNode, PauseThenStop) {
+  PauseThenStopNode::Register();
+
+  constexpr int num_pass = 2, num_batches = 10, batch_size = 1;
+  ASSERT_GE(num_pass, 2);
+  ASSERT_GT(num_batches, num_pass);
+  auto t_schema = schema({field("time", int32()), field("value", int32())});
+  ASSERT_OK_AND_ASSIGN(auto t_batches,
+                       MakeIntegerBatches({[](int row) -> int64_t { return row; },
+                                           [](int row) -> int64_t { return row + 1; }},
+                                          t_schema, num_batches, batch_size));
+
+  Declaration t_src = {
+      "source", SourceNodeOptions(t_batches.schema,
+                                  MakeDelayedGen(t_batches, "t_src", /*delay_sec=*/0.5,
+                                                 /*noisy=*/false))};
+  Declaration ctrl = {
+      PauseThenStopNode::kFactoryName, {t_src}, PauseThenStopNodeOptions(num_pass)};
+
+  ASSERT_OK_AND_ASSIGN(std::unique_ptr<RecordBatchReader> batch_reader,
+                       DeclarationToReader(ctrl, /*use_threads=*/false));
+
+  int64_t total_length = 0;
+  for (;;) {
+    ASSERT_OK_AND_ASSIGN(auto batch, batch_reader->Next());
+    if (!batch) {
+      break;
+    }
+    total_length += batch->num_rows();
+  }
+  ASSERT_EQ(static_cast<int64_t>(num_pass * batch_size), total_length);
```

Comment:
> Can we test this in a way that https://github.com/apache/arrow/blob/main/cpp/src/arrow/acero/exec_plan.h#L107 is actually called? (Since this what happens in real case).

Done.

Reply:
Maybe instead of using a thread pool and custom exec context, what if we 
(1) have a PauseNode/GateNode that calls PauseProducing as soon as it receives input
(2) expose a debug method that check if a SourceNode is paused. (SourceNode::IsPaused())
(3) use DeclarationToReader to start the plan
(4) Wait for SourceNode to be paused (using SourceNode::IsPaused())
(5) Call StopProducing on the created ExecPlan

I think the main difference from what Weston's above is the IsPaused() method, which might be useful in other back pressure related test as well

Reply:
Some more thoughts. I think we need
(1) A reliable/controlled way to trigger back pressure (similar to Weston's gate node idea)
(2) A way to validate that a Source node is paused because of backpresure (either via SourceNode::IsPaused()) or sth else
(3) A way to stop the execution while the query is running (either by providing an executor/thread to run the query while the main thread stop the execution, or have the main thread run the query and have a separate timed thread to stop the execution. Or just have one thread (if we manage to do this with just one thread)

I agree with Weston that this test sort of works, but since we are touching back pressure/cancellation quite a bit I think it is worthwhile to improve the testing around this.

Reply:
> I don't think it would help us to test whether a (source) node is paused. The current issue is about the pausing behavior, which manifests in backpressure, not about the pausing state, which may well just be a flag (that may even lack behavior, in an extreme case).

It is not perfect and relies on the assumption (if PauseProducing() is called on input node, then backpressure is properly applied). If we really want to check the source node is indeed paused, then I think we need to do what Weston suggest above which is perhaps a bit overkill. Checking the backpresure future or Checking PauseProducing() is called is a reasonable mid ground IMO.

> I don't think pausing is the object of testing here; backpressure behavior is, and pausing is just a tool to ensure backpressure.
Similar to above, yes it is not the object of testing, but is a proxy for back pressure since I am not aware of a way to check backpressure directly (the backpresure_future_ is a private variable)

> I can relate to this, though. If we're aiming for heavy testing of backpressure/cancellation interactions, we should probably create an issue for each kind of testing we have in mind. In my mind, the current PR deals with the current issue and has enough value on its own.
The test I have in mind is described above:
```
(1) have a PauseNode/GateNode that calls PauseProducing as soon as it receives input
(2) expose a debug method that check if a SourceNode is paused. (SourceNode::IsPaused())
(3) use DeclarationToReader to start the plan
(4) Wait for SourceNode to be paused (using SourceNode::IsPaused())
(5) Call StopProducing on the created ExecPlan
```
I agree this PR seems to fix the bug however I'd prefer to have a test that can reflect the actual usage, i.e. ExecPlan::StopProducing() gets called by user when back pressure is applied